# H6 DJ Ango

## Deploy Django 4 - Production Install 
28.2.2024 klo xx


(Karvinen 2021a)

## Django 4 Instant Customer Database Tutorial
28.2.2024 klo xx


(Karvinen 2021b)

## Käyttöympäristö

- MacOS Sonoma Versio 14.2.1 MacBookPro 13-tuumainen
- 2020 Prosessori: 1,4 GHz Neliytiminen Intel Core i5
- Näytönohjain: Intel Iris Plus Graphics 645, 1536 Mt
- Muisti: 8 Gt 2133 MHz LPDDR3
- Debian GNU/Linux 12 Bookworm

# Yksinkertainen esimerkkiohjelma Djangolla 
(27.2.2024 klo 14-16.45 ja 28.2.2024 klo 9.00-xx)

Aloitin lataamalla virtualenvin terminaalissa komennolla

    $ sudo apt-get install virtualenv

<img width="784" alt="Näyttökuva 2024-2-27 kello 10 54 08" src="https://github.com/HannaBurman/koulutyot/assets/156779955/4cb08e07-30ad-48e8-9fd2-0ecd067538dd">

Tämän jälkeen loin django kansion komennolla

    $ mkdir django

ja menin kansion sisälle komennolla ja loin virtuaaliympäristön Phyton 3:lla komennoilla

    $ cd django
    $ virtualenv --system-site-packages -p python3 env/

<img width="782" alt="Näyttökuva 2024-2-27 kello 14 41 42" src="https://github.com/HannaBurman/koulutyot/assets/156779955/6ce7c3e9-2d10-46aa-bb7e-a4ce3e2115a1">

Tämän jälkeen env kansio löytyy komennolla

    $ ls

<img width="291" alt="Näyttökuva 2024-2-27 kello 14 43 02" src="https://github.com/HannaBurman/koulutyot/assets/156779955/ee6ce2bb-2ad2-4f5c-aa95-b082ee4c3e50">

envi otetaan käyttöön komennolla (tarkoituksena käyttää pipiä envin sisältä)

    $ source env/bin/activate

<img width="516" alt="Näyttökuva 2024-2-27 kello 14 45 01" src="https://github.com/HannaBurman/koulutyot/assets/156779955/4aa33006-0804-46ba-8e6f-aabf3fd5a85e">

Seuraavalla komennolla nähdään, että pip asentuu env kansioon

    $ which pip 

<img width="421" alt="Näyttökuva 2024-2-27 kello 14 51 58" src="https://github.com/HannaBurman/koulutyot/assets/156779955/9888b002-5832-461e-8bd8-70a80f5f212b">

Seuraavaksi teen djangolle kansion jonne kirjoitan pelkän "django" (ja olen varovainen, etten tee kirjoitusvirheitä).

    $ micro requirements.txt

Komennolla 

    $ cat requirements.txt 

näen vielä suoraan terminaalista mitä tekstieditoriin on kirjoitettu. 

<img width="555" alt="Näyttökuva 2024-2-27 kello 15 00 06" src="https://github.com/HannaBurman/koulutyot/assets/156779955/55c0ff4d-cd42-4ca3-8043-97fd87dc7e17">

Seuraavaksi asennan djangon komennolla (asennus pipillä tapahtuu apt-getin ulkopuolelta)

    $ pip install -r requirements.txt

<img width="769" alt="Näyttökuva 2024-2-27 kello 15 01 41" src="https://github.com/HannaBurman/koulutyot/assets/156779955/4e54460b-f869-4048-9946-15164da0bd39">

Seuraavalla komennolla saan esiin kaikki djangon komennot

    $ django-admin

<img width="438" alt="Näyttökuva 2024-2-27 kello 15 05 26" src="https://github.com/HannaBurman/koulutyot/assets/156779955/0993d5df-ccf4-4295-a3b5-b6d640a33070">

Tein projektilleni uuden kansion komennolla

    $ django-admin startproject hannacom

Tein testipalvelimen paikalliseen testaukseen komennolla

    $ ./manage.py runserver #DO NOT EXPOSE TO INTERNET

<img width="783" alt="Näyttökuva 2024-2-27 kello 15 14 01" src="https://github.com/HannaBurman/koulutyot/assets/156779955/69de6cce-b6e5-49c4-8eaa-de78916f79bd">

Sivusto näyttää tältä

<img width="782" alt="Näyttökuva 2024-2-27 kello 14 23 26" src="https://github.com/HannaBurman/koulutyot/assets/156779955/adf20b8c-c47b-4612-b87f-7ad8cb952bf8">

Ctrl + c testisivu tapetaan, jolloin testisaitti ei enää toimi. 

<img width="704" alt="Näyttökuva 2024-2-27 kello 14 18 01" src="https://github.com/HannaBurman/koulutyot/assets/156779955/af219200-b2ae-4613-a08b-bf36233acf20">

Seuraavaksi päivitän tietokannat komennoilla

    $ ./manage.py makemigrations
    $ ./manage.py migrate
    
<img width="801" alt="Näyttökuva 2024-2-28 kello 9 35 54" src="https://github.com/HannaBurman/koulutyot/assets/156779955/31b509e2-bce7-4fb7-80b3-1cb482d750ef">

Sitten lisään käyttäjän. Uuden käyttäjän luonnin aloitan asentamalla pwgen -työkalun, jolla saan generoituja salasanoja komennolla

    $ sudo apt-get install pwgen

Seuraavalla komennolla luon 20 merkkiä sisältävän salasanan

    $ pwgen -s 20 1 # randomize a password

<img width="765" alt="Näyttökuva 2024-2-28 kello 9 54 13" src="https://github.com/HannaBurman/koulutyot/assets/156779955/245c7c66-32d9-4328-9c0d-ce672d034775">

Sitten käyttäjän luontiin komennolla 

  $ ./manage.py createsuperuser

<img width="686" alt="Näyttökuva 2024-2-28 kello 9 53 10" src="https://github.com/HannaBurman/koulutyot/assets/156779955/a217cb27-5664-4582-baff-5b89859dd524">

Saadakseni serverin käynttiin, annoin uudestaan komennon 

    ./manage.py runserver #DO NOT EXPOSE TO INTERNET

<img width="800" alt="Näyttökuva 2024-2-28 kello 10 03 08" src="https://github.com/HannaBurman/koulutyot/assets/156779955/6a3b11e3-87b7-401a-a118-a1c0b6c7fa10">

Lisäsäin osoitekentän loppuun /admin/ annoin salasanan ja pääsin sisälle

<img width="796" alt="Näyttökuva 2024-2-28 kello 9 57 28" src="https://github.com/HannaBurman/koulutyot/assets/156779955/18b25531-9bd2-471c-86d2-1cbc61d3bb0b">

Tämän jälkeen terminaalissa näytti seuraavalta

<img width="800" alt="Näyttökuva 2024-2-28 kello 10 04 44" src="https://github.com/HannaBurman/koulutyot/assets/156779955/6cb93528-850c-4e9d-8a21-f25210b78220">

Nyt vasta huomasin, että kelloni jälleen kerran on väärässä ajassa, joten korjasin sen komennolla

    $ sudo apt-get install ntp
    $ sudo service ntp start
    $ sudo systemctl start ntp

Seuraavaksi tein vielä huvin vuoksi edelliset, jotta näen tulokset oikeassa ajassa. 

<img width="800" alt="Näyttökuva 2024-2-28 kello 10 09 51" src="https://github.com/HannaBurman/koulutyot/assets/156779955/624f4c50-97b3-4be2-90de-f7a3afa8e30a">

Viikkotehtävän ohjeissa oli seuraavaksi luvassa uuden käyttäjän luonti. Uudesta käyttäjästä on tarkoitus tehdä staff- ja superuser, jotta hän voi muokata mitä tahansa tietoa verkkohallintaliittymän kautta. Jälleen tärkeää käyttää vahvaa salasanaa.

Aloitin luomisen antamalla komennon 

    $ ./manage.py startapp crm

Tämä loi uuden kansion nimeltään crm/ asiakassuhteiden hallinta -sovellukselleni.

Lisäsin sovelluksen INSTALLED_APPS-kohtaan settings.py-tiedostossa komennolla

    $ micro hannacom/settings.py

Ja lisäsin listaan 'crm' ja tallensin. 
    
<img width="400" alt="Näyttökuva 2024-2-28 kello 10 21 58" src="https://github.com/HannaBurman/koulutyot/assets/156779955/9625bf00-b5dc-4303-9712-95e105daa479">

Seuraavaksi siirryn crm hakemistoon muokatakseni tiedostoa models.py komennolla   
    
    $ micro crm/models.py 

<img width="493" alt="Näyttökuva 2024-2-28 kello 10 28 03" src="https://github.com/HannaBurman/koulutyot/assets/156779955/09e11cdb-b7a1-4438-a0b7-253bcb6e3b76">

Seuraavaksi päivitän tietokannat komennoilla

    $ ./manage.py makemigrations
    $ ./manage.py migrate    

Herjasi kuitenkin kirjoitusvirhettä, joten korjasin sen ja annoin komennon uudelleen. 

<img width="790" alt="Näyttökuva 2024-2-28 kello 10 30 20" src="https://github.com/HannaBurman/koulutyot/assets/156779955/746dfbfb-6ffc-4055-86cd-aa0370653799">

Nyt onnistui

<img width="644" alt="Näyttökuva 2024-2-28 kello 10 35 04" src="https://github.com/HannaBurman/koulutyot/assets/156779955/09686381-808e-4057-b2bb-9968979670c2">

Jotta uusi tietokanta tulisi adminissa näkyviin, pitää se rekisteröidä komennolla

    $ micro crm/admin.py

<img width="412" alt="Näyttökuva 2024-2-28 kello 10 37 59" src="https://github.com/HannaBurman/koulutyot/assets/156779955/3a4afdc0-2854-4845-b666-6af133ab0004">

Käynnistin jälleen serverin komennolla

    $ ./manage.py runserver

Menin sivulle http://127.0.0.1:8000/admin/ ja se toimii. 

<img width="703" alt="Näyttökuva 2024-2-28 kello 10 40 24" src="https://github.com/HannaBurman/koulutyot/assets/156779955/717d0629-5abe-48ed-804b-57420dd04402">

Lisäsin sivustolla kaksi asiakasta. 

<img width="703" alt="Näyttökuva 2024-2-28 kello 10 45 04" src="https://github.com/HannaBurman/koulutyot/assets/156779955/cf6e8136-10c3-4a27-b354-760ff5d6b903">

Koska haluan hienot nimet näkyviin, annoin terminaalissa komennon

    $ micro crm/models.py

Huomasin typon, jonka samalla korjasin (nsme = name) ja annoin uudelleen komennot

    $ ./manage.py makemigrations
    $ ./manage.py migrate  

<img width="509" alt="Näyttökuva 2024-2-28 kello 10 51 15" src="https://github.com/HannaBurman/koulutyot/assets/156779955/25748a4a-6330-4c90-af29-68543b3966f3">



    





Kirjaudu ulos verkkohallinnasta ja kirjaudu sisään uutena käyttäjänäsi.

Voitko nyt muokata tietoja? Hienoa, sinulla on nyt uusi verkkohallinnan käyttäjä.



Voit käyttää testipalvelinta, kunhan se ei näy Internetiin.
Riittää, kun ohjelmasi näkyy esimerkiksi Django Adminsissa.
Voit halutessasi tehdä aivan samanlaisen kuin Teron CRM-esimerkissä.

# Djangon tuotantotyyppinen asennus
(28.2.2024 klo xx)

Voit tehdä asennuksen omalle, paikalliselle virtuaalikoneelle. Sen ei tarvitse näkyä Internetiin.

Vinkit

Käytä systemaattista työtapaa
Tee muistiinpanot samalla kun työskentelet
Lue virheilmoitukset lokeista
Pienin testattava kokonaisuus kerralla

## Lähteet

- Karvinen,  T. 2021a. Deploy Django 4 - Production Install. Luettavissa: https://terokarvinen.com/2022/deploy-django/. Luettu 28.2.2024
- Karvinen, T. 2021b. Django 4 Instant Customer Database Tutorial. Luettavissa: https://terokarvinen.com/2022/django-instant-crm-tutorial/. Luettu 28.2.2024.

